### ISSUE TITLE
Flutter cannot build against ARM simulators (on ARM macOS host)

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#64502]
Targeting an ARM simulator fails:
```
[plugin_lint_mac] [STDOUT] stdout:     ld: building for iOS Simulator, but linking in dylib built for iOS, file '/var/folders/3y/l0z1x3693dl_8n0qybp4dqwh0000gn/T/CocoaPods-Lint-20200812-67637-y2vglb-test_plugin_objc/Pods/Flutter/Flutter.framework/Flutter' for architecture arm64
[plugin_lint_mac] [STDOUT] stdout:     clang: error: linker command failed with exit code 1 (use -v to see invocation
```

Flutter.framework has 2 arm iOS slices and 1 x86 simulator slice.  There is no ARM simulator slice.

Also seen on an x86 host with iOS 14 simulators (so Xcode 12) when `ONLY_ACTIVE_ARCH=NO` (configuration=Release)

Blocked by https://github.com/flutter/flutter/issues/66926

**Comment by jmagman** [comment]
Not sure if there's any dart work to do, since they already support arm simulators?  This may just require building the Flutter.framework with a new architecture slice.

**Comment by jmagman** [comment]
See https://github.com/flutter/flutter/issues/2089, hopefully this is much less work than that was.

**Comment by jmagman** [comment]
Lots of spots where `simulator` is synonymous with x86 https://github.com/flutter/engine/blob/ff62dec5c96d3560b36ca0a3924d28a3c97b7788/sky/tools/create_ios_framework.py#L32

**Comment by jmagman** [comment]
Had to turn off linters when I upgrade Cirrus to Xcode 12 https://github.com/flutter/flutter/pull/63252
Plugins CI is going to hit the same failure when they try to upgrade.

**Comment by alexmarkley** [comment]
I'm encountering this problem as well. (Actually opened a duplicate issue, oops.)

Any update on this? I'm happy to help if I can.

**Comment by alexmarkley** [comment]
@jmagman In the context of some of the plumbing / tooling tasks that need to be done here, I would be happy to help, but I'm currently stuck on this issue: https://github.com/dart-lang/sdk/issues/43488

If you've got a line to the Dart team (or an idea for circumventing this blocker some other way) I would appreciate the help.

**Comment by Poweranimal** [comment]
There's a workaround for this issue:
Add `config.build_settings['ONLY_ACTIVE_ARCH'] = 'YES' ` to the Podfile's post_install.
E.g:
`post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings['ENABLE_BITCODE'] = 'NO'
      # Workaround for https://github.com/flutter/flutter/issues/64502
      config.build_settings['ONLY_ACTIVE_ARCH'] = 'YES'
    end
  end
end
`

I found the solution in another related issue here: https://github.com/flurry/flurry-ios-sdk/issues/207#issuecomment-696337894

**Comment by jmagman** [comment]
> config.build_settings['ONLY_ACTIVE_ARCH'] = 'YES'

That will let you build, but when you archive it won't build a fat binary with all the architectures your app supports. And it's only relevant if you're using plugins.

**Comment by jmagman** [comment]
Fixing this will also let us turn back on plugin linters, and will let the plugins repo update Cirrus to Xcode 12.
https://github.com/flutter/flutter/pull/63252#issuecomment-679335036
https://github.com/flutter/flutter/blob/d693d2b9d282c30ef14b444a7860c46893499f7d/dev/devicelab/bin/tasks/plugin_lint_mac.dart#L52-L53

**Comment by jmagman** [comment]
Also, dependent on the builders using iOS 14 https://github.com/flutter/flutter/issues/65901.

**Comment by hiphonezhu** [comment]
how to resolve?

**Comment by jmagman** [comment]
This will require https://github.com/flutter/flutter/issues/60109 since the arm64 `iphoneos` and `iphonesimulator` slices can no longer be lipo'd together.

**Comment by bouraine** [comment]
this could help (see Ayan Sengupta answer): 
-[stackoverflow 63607158]( https://stackoverflow.com/questions/63607158/xcode-12-building-for-ios-simulator-but-linking-in-object-file-built-for-ios)

**Comment by jmagman** [comment]
This issue is tracking Flutter shipping ARM simulator binaries, not working around the lack thereof.

**Comment by deadsoul44** [comment]
Is there any ETA? I am planning to buy a new M1 Mac Mini? Thank you.

**Comment by joniaiuser** [comment]
I was able to build my app successfully on M1 Macbook Air running Terminal in Rosetta simulation against a iOS iPhone 12 Pro **Simulator**, Flutter stable version. I have loads of packages and all and it still worked. I followed the instructions here: https://github.com/flutter/flutter/wiki/Developing-with-Flutter-on-Apple-Silicon

However, I ran into this issue where this error that used to appear sporadically is now flooding my terminal and basically spamming non-stop.

![image](https://user-images.githubusercontent.com/64962001/100514197-a7d83600-31ad-11eb-9f54-2520dcb5104e.png)

**Comment by yzxben** [comment]
> I was able to build my app successfully on M1 Macbook Air running Terminal in Rosetta simulation against a iOS iPhone 12 Pro **Simulator**, Flutter **stable** version. I have loads of packages and all and it still worked. I followed the instructions here: https://github.com/flutter/flutter/wiki/Developing-with-Flutter-on-Apple-Silicon

@joniaiuser I suspect you are not deploying it to the simulator (eg running it from XCode)

**Comment by joniaiuser** [comment]
@yzxben  I was running from the command line i.e. flutter run.. is that not deploying to simulator? Ironically I couldn't get dev build to work on my actual device.. release mode seems fine

**Comment by BernardGatt** [comment]
Just tested this. If you run XCode & Simulator on Apple architecture it's not possible to build and test on simulator as described in the wiki.

But, if you run both XCode & Simulator using rosetta (Right Click -> Get Info -> Open using Rosetta) it will build and run on simulator. I'll switch to this workflow tomorrow to see if I find any problems.

I'm using an Add2App integration btw.

**Comment by acoutts** [comment]
Potentially interesting data point - while I can't target ARM simulators, I am able to target my mac itself and build the app to run on macOS as an iOS app.

<img width="293" alt="Screen Shot 2020-12-02 at 5 23 12 PM" src="https://user-images.githubusercontent.com/9597207/100938649-10335a00-34c3-11eb-94ff-ba0983b3a6c1.png">

Note: only from within Xcode, won't have Flutter hot reload or anything.

**Comment by jmagman** [comment]
@acoutts We're tracking with https://github.com/flutter/flutter/issues/60113, see also https://github.com/flutter/flutter/issues/66184.

**Comment by PandaGeek1024** [comment]
I'm getting a similar issue with multiple flavors. Building an app with dev and prod flavors. Android works fine but ios is facing a bit of trouble.

What I did as follows:
1. created two dev and prod schemes and two targets respectively.
2. add debug and release config files for different flavors
3. add plugin 'flutter_pdfview' in flutter and run pub get.

Things work fine when run through terminals
`flutter run --flavor prod` or `flutter run --flavor prod`

But when I build it through xcode, I'm getting this error
```
Ld /Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Intermediates.noindex/Pods.build/Debug-dev-iphonesimulator/flutter_pdfview.build/Objects-normal/arm64/Binary/flutter_pdfview normal arm64 (in target 'flutter_pdfview' from project 'Pods')
    cd /Users/pengfeihuang/Developer/Flutter/projects/flavors/ios/Pods
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target arm64-apple-ios9.0-simulator -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.0.sdk -L/Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Products/Debug-dev-iphonesimulator/flutter_pdfview -F/Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Products/Debug-dev-iphonesimulator/flutter_pdfview -F/Users/pengfeihuang/Developer/Flutter/projects/flavors/ios/Pods/../Flutter -filelist /Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Intermediates.noindex/Pods.build/Debug-dev-iphonesimulator/flutter_pdfview.build/Objects-normal/arm64/flutter_pdfview.LinkFileList -install_name @rpath/flutter_pdfview.framework/flutter_pdfview -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -dead_strip -Xlinker -object_path_lto -Xlinker /Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Intermediates.noindex/Pods.build/Debug-dev-iphonesimulator/flutter_pdfview.build/Objects-normal/arm64/flutter_pdfview_lto.o -Xlinker -objc_abi_version -Xlinker 2 -fobjc-arc -fobjc-link-runtime -framework Flutter -framework Foundation -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Intermediates.noindex/Pods.build/Debug-dev-iphonesimulator/flutter_pdfview.build/Objects-normal/arm64/flutter_pdfview_dependency_info.dat -o /Users/pengfeihuang/Library/Developer/Xcode/DerivedData/Runner-ducgzqqypwgexiekvbigdedzqhbw/Build/Intermediates.noindex/Pods.build/Debug-dev-iphonesimulator/flutter_pdfview.build/Objects-normal/arm64/Binary/flutter_pdfview

ld: building for iOS Simulator, but linking in dylib built for iOS, file '/Users/pengfeihuang/Developer/Flutter/projects/flavors/ios/Flutter/Flutter.framework/Flutter' for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I have created a sample app [here](https://github.com/PandaGeek1024/flavors).



```
[✓] Flutter (Channel stable, 1.22.5, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-NZ)
    • Flutter version 1.22.5 at /Users/pengfeihuang/fvm/versions/stable
    • Framework revision 7891006299 (4 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/pengfeihuang/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = /Users/pengfeihuang/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.10.0

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin installed
    • Dart plugin version 201.9317
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.0

[✓] Connected device (1 available)
    • iPhone SE (2nd generation) (mobile) • 01768AA3-066A-483C-BAED-91EDC3E50665 • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)

• No issues found!

```

**Comment by jmagman** [comment]
@PandaGeek1024 Yup that's what this issue is tracking.  Plus [Flutter isn't really supported on the simulator in release mode](https://github.com/flutter/flutter/issues/11754) so you probably aren't testing what you think you're testing in the simulator.

**Comment by PandaGeek1024** [comment]
@jmagman If you are refering to the prod flavor, sure Im not expecting it to be a real release mode. 

But is there an explanation why it works with flutter command line but not with xcode? This is confusing.

**Comment by jmagman** [comment]
@PandaGeek1024 I'm not sure, I wouldn't expect it to work for either.  Can you file a new issue and include a full `flutter run -v` output where it works, as well as the full Xcode logs when it doesn't?

**Comment by jmagman** [comment]
Nevermind @PandaGeek1024 you don't need to file a new issue, I remember why this works from `flutter` but not from Xcode--the command line tool always calls `-arch x86_64` for the simulator.
https://github.com/flutter/flutter/blob/74d58105d1563e130ff196eac0f195def8214b10/packages/flutter_tools/lib/src/ios/mac.dart#L253

**Comment by jmagman** [comment]
This issue is tracking actually getting the ARM bits into the Flutter engine.  However to work around it until that's done, you can explicitly exclude arm64 from your simulator archs.  I tried to do this programmatically for all projects in https://github.com/flutter/flutter/pull/73458 but unfortunately versions of Xcode older than 12 do not handle the dance between `EXCLUDED_ARCHS` and `VALID_ARCHS` correctly.  The minimum required version of Xcode is currently Xcode 11, and we can't break users with that version installed.

To work around manually if you are on Xcode 12+:
1. If you are using plugins, add these line to your `ios/Podfile`.  If this file doesn't exist in your project, skip this step.
```ruby
post_install do |installer|
  installer.pods_project.targets.each do |target|
    flutter_additional_ios_build_settings(target)

    # Add the following:
    target.build_configurations.each do |build_configuration|
      build_configuration.build_settings['EXCLUDED_ARCHS[sdk=iphonesimulator*]'] = 'arm64 i386'
    end
  end
end
```
2. Open `ios/Runner.xcworkspace` in Xcode.  In your Runner target, find the Excluded Architectures (`EXCLUDED_ARCHS`) build setting. Click the right arrow disclosure indicator icon to expand the available build configurations. Hover over Debug and click the plus icon. Change **Any SDK** to **Any iOS Simulator SDK**. Add arm64 and i386 to the build settings value.
<img width="1142" alt="Screen Shot 2021-01-12 at 6 37 29 PM" src="https://user-images.githubusercontent.com/682784/104399359-58538c80-5505-11eb-9e91-23c6cde7f416.png">

You should now be able to run your app in the simulator from Xcode.

**Comment by bhanuka96** [comment]
same issue

**Comment by gourav6m17** [comment]
Hey everyone!
I'm facing a problem while I'm run my app on iOS simulator but for android simulator emulator it is working fine.
It throws me error as below:
`    /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.1/lib/cocoapods/command/install.rb:52:in `run'
    /Library/Ruby/Gems/2.6.0/gems/claide-1.0.3/lib/claide/command.rb:334:in `run'
    /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.1/lib/cocoapods/command.rb:52:in `run'
    /Library/Ruby/Gems/2.6.0/gems/cocoapods-1.10.1/bin/pod:55:in `<top (required)>'
    /usr/local/bin/pod:23:in `load'
    /usr/local/bin/pod:23:in `<main>'
    ```
    ――― TEMPLATE END ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
    [!] Oh no, an error occurred.
    Search for existing GitHub issues similar to yours:
    https://github.com/CocoaPods/CocoaPods/search?q=dlsym%280x7fad0ef3a210%2C+Init_ffi_c%29%3A+symbol+not+found+-+%2FLibrary%2FRuby%2FGems%2F2.6.0%2Fgems%2Fffi-1.14.2%2Flib%2Fffi_c.bundle&type=Issues
    If none exists, create a ticket, with the template displayed above, on:
    https://github.com/CocoaPods/CocoaPods/issues/new
    Be sure to first read the contributing guide for details on how to properly submit a ticket:
    https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md
    Don't forget to anonymize any private data!
    Looking for related issues on cocoapods/cocoapods...
    Found no similar issues. To create a new issue, please visit:
    https://github.com/cocoapods/cocoapods/issues/new
Error output from CocoaPods:
↳
    [!] Automatically assigning platform `iOS` with version `8.0` on target `Runner` because no platform was specified. Please specify a platform for this target in your Podfile. See `https://guides.cocoapods.org/syntax/podfile.html#platform`.
Error running pod install
Error launching application on iPhone 12 Pro Max.
Exited (sigterm).    `

**Comment by jmagman** [comment]
@gourav6m17 CocoaPods is working on this https://github.com/CocoaPods/CocoaPods/issues/9907

**Comment by jmagman** [comment]
> However to work around it until that's done, you can explicitly exclude arm64 from your simulator archs.  I tried to do this programmatically for all projects in https://github.com/flutter/flutter/pull/73458 but unfortunately versions of Xcode older than 12 do not handle the dance between `EXCLUDED_ARCHS` and `VALID_ARCHS` correctly. 

This has been fixed with https://github.com/flutter/flutter/pull/73828.  On that PR you don't need to set `EXCLUDED_ARCHS` in your Podfile.

**Comment by jigarfumakiya** [comment]
@jmagman 
Any update or workaround for this issue?

**Comment by jmagman** [comment]
@jigarfumakiya yes, this is worked around in 2.1.0-12.0.pre via #73458 which is currently available in the [beta channel](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos).  See https://github.com/flutter/flutter/issues/64502#issuecomment-759165857 for the workaround if you can't upgrade to the beta channel.

**Comment by firatcetiner** [comment]
Is there any workaround for this issue? I'm facing this issue with Flutter 2.2.2 and it seems there is no way to build to iOS simulator on my Mac device.

### flutter doctor
```
[✓] Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale en-TR)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.57.1)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by jans-y** [comment]
@firatcetiner 
Just run the Xcode.app and Simulator.app under Rosetta.

You will find Xcode under your Applications folder and Simulator inside the Xcode.app contents or just open this path:
/Applications/Xcode.app/Contents/Developer/Applications/Simulator.app

Helpful links for beginners with macOS:
https://support.apple.com/en-us/HT211861
https://www.macobserver.com/tmo/article/how-to-inspect-an-os-x-package-file-pkg-without-installing

**Comment by firatcetiner** [comment]
> @firatcetiner
> Just run the Xcode.app and Simulator.app under Rosetta.
> 
> You will find Xcode under your Applications folder and Simulator inside the Xcode.app contents or just open this path:
> /Applications/Xcode.app/Contents/Developer/Applications/Simulator.app
> 
> Helpful links for beginners with macOS:
> https://support.apple.com/en-us/HT211861
> https://www.macobserver.com/tmo/article/how-to-inspect-an-os-x-package-file-pkg-without-installing

I'm already using a device with an Intel processor, so building under Rosetta is not an option for me.

**Comment by jans-y** [comment]
@firatcetiner
I misunderstood your problem then. I believe this issue has been created for the ARM Macs (M1) – "ARM macOS host".

**Comment by jmagman** [comment]
@firatcetiner Please file a new issue and completely fill out the template.  Include the `--verbose` output of a `flutter run` or `flutter build ios` so I can see the Xcode build logs. This issue is tracking building native arm64 iOS simulator apps.

> Just run the Xcode.app and Simulator.app under Rosetta.

@jans-y You shouldn't need to do that, a Flutter app will run on the native Simulator.  If you're unable to do so, then that's unexpected, please file a new GitHub issue with verbose logging. Suggesting things like that in a GitHub issue prevents us from tracking down the problems people are seeing since they shouldn't need to apply an undocumented workaround.
